sign/slhdsa/wotsp.go

```diff
@@ -0,0 +1,125 @@
+package slhdsa
+
+// See FIPS 205 -- Section 5
+// Winternitz One-Time Signature Plus Scheme
+
+const (
+	wotsW    = 16 // wotsW is w = 2^lg_w, where lg_w = 4.
+	wotsLen2 = 3  // wotsLen2 is len_2 fixed to 3.
+)
+
+type (
+	wotsPublicKey []byte // n bytes
+	wotsSignature []byte // wotsLen()*n bytes
+)
+
+func (p *params) wotsSigSize() uint32 { return p.wotsLen() * p.n }
+func (p *params) wotsLen() uint32     { return p.wotsLen1() + wotsLen2 }
+func (p *params) wotsLen1() uint32    { return 2 * p.n }
+
+func (ws *wotsSignature) fromBytes(p *params, c *cursor) {
+	*ws = c.Next(p.wotsSigSize())
+}
+
+// See FIPS 205 -- Section 5 -- Algorithm 5.
+func (s *state) chain(x []byte, index, step uint32, addr address) (out []byte) {
+	out = x
+	s.F.address.Set(addr)
+	for j := index; j < index+step; j++ {
+		s.F.address.SetHashAddress(j)
+		s.F.SetMessage(out)
+		out = s.F.Final()
+	}
+	return
+}
+
+// See FIPS 205 -- Section 5.1 -- Algorithm 6.
+func (s *statePriv) wotsPkGen(addr address) wotsPublicKey {
+	s.PRF.address.Set(addr)
+	s.PRF.address.SetTypeAndClear(addressWotsPrf)
+	s.PRF.address.SetKeyPairAddress(addr.GetKeyPairAddress())
+
+	s.T.address.Set(addr)
+	s.T.address.SetTypeAndClear(addressWotsPk)
+	s.T.address.SetKeyPairAddress(addr.GetKeyPairAddress())
+
+	s.T.Reset()
+	wotsLen := s.wotsLen()
+	for i := uint32(0); i < wotsLen; i++ {
+		s.PRF.address.SetChainAddress(i)
+		sk := s.PRF.Final()
+
+		addr.SetChainAddress(i)
+		tmpi := s.chain(sk, 0, wotsW-1, addr)
+
+		s.T.WriteMessage(tmpi)
+	}
+
+	return s.T.Final()
+}
+
+// See FIPS 205 -- Section 5.2 -- Algorithm 7.
+func (s *statePriv) wotsSign(sig wotsSignature, msg []byte, addr address) {
```

Comment:
Conventionally, passing the output in the first parameter when memory is pre-allocated, instead of returning it as the output of the function.
Besides, passing a slice in Go is always writable.